src/bundle-manifest.ts

```diff
@@ -166,54 +209,28 @@ export function generateSharedDepsMergeStrategy(minEntrypoints?: number):
   if (minEntrypoints === undefined) {
     minEntrypoints = 2;
   }
-  return (bundles: Bundle[]): Bundle[] => {
-    const newBundles: Bundle[] = [];
-    const sharedBundles: Bundle[] = [];
-    const allEntrypoints = new Set<UrlString>();
-    for (const bundle of bundles) {
-      bundle.entrypoints.forEach(
-          (entrypoint) => allEntrypoints.add(entrypoint));
-      if (bundle.entrypoints.size >= minEntrypoints &&
-          !getBundleEntrypoint(bundle)) {
-        sharedBundles.push(bundle);
-      } else {
-        newBundles.push(
-            new Bundle(new Set(bundle.entrypoints), new Set(bundle.files)));
-      }
-    }
-    for (let i = 0; i < newBundles.length; i++) {
-      const bundle = newBundles[i];
-      if (setEquals(bundle.entrypoints, allEntrypoints)) {
-        newBundles.splice(i, 1)[0];
-        sharedBundles.push(bundle);
-        break;
-      }
-    }
-    if (sharedBundles.length > 0) {
-      newBundles.push(mergeBundles(sharedBundles));
-    }
-    return newBundles;
-  };
+  return (bundles: Bundle[]): Bundle[] => mergeMatchingBundles(
+             bundles,
+             bundle => bundle.entrypoints.size >= minEntrypoints &&
+                 !getBundleEntrypoint(bundle));
 }
 
 /**
  * A bundle strategy function which merges all shared dependencies into a
  * bundle for an application shell.
  */
 export function generateShellMergeStrategy(
-    shell: UrlString, minEntrypoints: number): BundleStrategy {
-  return (bundles: Bundle[]): Bundle[] => {
-    const newBundles = generateSharedDepsMergeStrategy(minEntrypoints)(bundles);
-    const shellBundle = newBundles.find((bundle) => bundle.files.has(shell));
-    const sharedBundle =
-        newBundles.find((bundle) => bundle.entrypoints.size > 1);
-    if (shellBundle && sharedBundle && shellBundle !== sharedBundle) {
-      newBundles.splice(newBundles.indexOf(shellBundle), 1);
-      newBundles.splice(newBundles.indexOf(sharedBundle), 1);
-      newBundles.push(mergeBundles([shellBundle, sharedBundle]));
-    }
-    return newBundles;
-  };
+    shell: UrlString, minEntrypoints?: number): BundleStrategy {
+  if (minEntrypoints === undefined) {
+    minEntrypoints = 2;
+  }
+  return composeStrategies([
+    generateEagerMergeStrategy(shell),
+    generateSharedDepsMergeStrategy(minEntrypoints),
+    generateMatchMergeStrategy(
```

Comment:
Per discussion earlier, it turns out that the generateSharedDepsMergeStrategy is actually the redundant one.  We need the generateMatchMergeStrategy because that's what the expectation is for shop-- there's all these shared dependencies which should be inlined into the shell but none of them are transitive dependencies of shell, and the match merge is the one that does that.